### PR TITLE
GH-33559: [C++] Move select-k implementation into separate module

### LINKED ISSUE #33559: [C++] Move select-k implementation into separate module
The select-k kernel implementations are currently in `vector_sort.cc`, amongst other things.
To make the code more readable and faster to compiler, we should move them into their own file.

**Reporter**: [Antoine Pitrou](https://issues.apache.org/jira/browse/ARROW-18395) / @pitrou
**Assignee**: [Ben Harkins](https://issues.apache.org/jira/browse/ARROW-18395) / @benibus
#### PRs and other links:
- [GitHub Pull Request #14753](https://github.com/apache/arrow/pull/14753)

<sub>**Note**: *This issue was originally created as [ARROW-18395](https://issues.apache.org/jira/browse/ARROW-18395). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by benibus**
See: [ARROW-18395](https://issues.apache.org/jira/browse/ARROW-18395)
* Closes: #33559

### CODE REVIEW

**Comment by benibus** [comment]
Not sure if there are any strong opinions about the inline functions in `vector_sort_internal.h`... Some of the definitions could remain where they were if necessary.

Also, I removed a bunch of redundant `#include`s in the implementations - although I don't really know if they were just cruft or meant to prevent future breakages, as quite a few things are being implicitly pulled in by two headers.

File: cpp/src/arrow/compute/kernels/vector_sort_internal.h [hunk 1]
```diff
@@ -456,6 +458,402 @@ Status SortChunkedArray(ExecContext* ctx, uint64_t* indices_begin, uint64_t* ind
...
+  }
+
+  const std::vector<ResolvedSortKey>& sort_keys_;
+  const NullPlacement null_placement_;
+  std::vector<std::unique_ptr<ColumnComparator<ResolvedSortKey>>> column_comparators_;
+  Status status_;
+};
+
+// Sort a batch using a single sort and multiple-key comparisons.
+class MultipleKeyRecordBatchSorter : public TypeVisitor {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_sort_internal.h:708]: I'm not sure it's useful to expose this internal class in a `.h` file? By the looks of it, only `MultipleKeyRecordBatchSorter::ResolvedSortKey` is used by the select-k kernels?

**Review by pitrou — COMMENTED** [review]
Thanks a lot. A few simple comments.

**Comment by pitrou** [comment]
@benibus Are you willing to move forward on this?

**Comment by benibus** [comment]
@pitrou Yes. In fact, I'm working on it right now.

File: cpp/src/arrow/compute/kernels/vector_sort_internal.h [hunk 2]
```diff
@@ -456,6 +458,402 @@ Status SortChunkedArray(ExecContext* ctx, uint64_t* indices_begin, uint64_t* ind
...
+    const TableOrBatch& table_or_batch, const std::vector<SortKey>& sort_keys) {
+  return ResolveSortKeys<ResolvedSortKey>(
+      *table_or_batch.schema(), sort_keys, [&](const SortField& f) {
+        return ResolvedSortKey{table_or_batch.column(f.field_index), f.order};
+      });
+}
+
+// Returns nullptr if no column matching `ref` is found, or if the FieldRef is
+// a nested reference.
+inline std::shared_ptr<ChunkedArray> GetTableColumn(const Table& table,
```
[the remark this answers is not in the document]
Comment (benibus) [reply cpp/src/arrow/compute/kernels/vector_sort_internal.h:507]: Agreed. I just opened an issue for it: https://github.com/apache/arrow/issues/14939

Do you think that should be done first?

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/vector_sort_internal.h:507]: I think you could just do it in this PR.

**Comment by benibus** [comment]
Not quite sure if you think exposing the one common `ResolvedSortKey` in the header would be worthwhile - since I basically just copied it into select-k's `RecordBatchSelector`. In practice, there are 3 unique `ResolvedSortKey` definitions across 4 classes.

File: cpp/src/arrow/compute/kernels/vector_sort_internal.h [hunk 3]
```diff
@@ -456,6 +457,243 @@ Status SortChunkedArray(ExecContext* ctx, uint64_t* indices_begin, uint64_t* ind
...
+  }
+
+  const std::vector<ResolvedSortKey>& sort_keys_;
+  const NullPlacement null_placement_;
+  std::vector<std::unique_ptr<ColumnComparator<ResolvedSortKey>>> column_comparators_;
+  Status status_;
+};
+
+inline Result<std::shared_ptr<ArrayData>> MakeMutableUInt64Array(
+    std::shared_ptr<DataType> out_type, int64_t length, MemoryPool* memory_pool) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_sort_internal.h:691]: Is `out_type` unused here?

**Comment by pitrou** [comment]
> Not quite sure if you think exposing the one common `ResolvedSortKey` in the header would be worthwhile - since I basically just copied it into select-k's `RecordBatchSelector`. In practice, there are 3 unique `ResolvedSortKey` definitions across 4 classes.

Basically there should be two possible `ResolvedSortKey` implementations: one for record batch (contiguous) input, the other for table (chunked) input. So, yes, I think those two implementations should be put in the header.

**Comment by benibus** [comment]
Probably worth noting that `GetTableColumn` (which I replaced in the last commit) isn't exactly equivalent to `FieldRef::GetOne` since the former will (a) reject nested fields and (b) won't search for multiple matching `FieldRef`s - unlike the latter, which calls `FieldRef::FindAll` under the hood.

That being said, those constraints/optimizations don't currently apply to the `RecordBatch` functions, so I'm not quite sure what the call would be here.

**Comment by pitrou** [comment]
Interesting. I don't think we intended to support nested fields (we would have to decide whether to take into account outer-level nulls or not), so we should perhaps disallow them everywhere for now.

**Comment by benibus** [comment]
Yeah, makes sense. I can save the FieldRef/Path stuff for a later date.

> we would have to decide whether to take into account outer-level nulls or not

That was a point of uncertainty for me as well. My implementation of `FieldPath::FindAll` for `Table`s has child fields inherit their parent's validity via flattening, but the existing versions just use their underlying `ArrayData` directly (most likely to avoid allocating a new `Array`?), which would raise even more questions.

**Comment by benibus** [comment]
@pitrou After looking deeper, what I said earlier isn't entirely correctly. Nested fields are only accepted in a few cases where `FindSortKeys` isn't called.  I tightened those up in the latest commit. There were also some inconsistencies in error reporting for missing/invalid `FieldRef`s (particularly in the select-k module), so I attempted to fix those as well. 

Also, due to the error message changes, one of the python tests is failing in the appveyor build. I figured I'd wait for your opinion before fixing it - but I suspect you'd agree that the new error diagnostics would be more informative.

**Comment by pitrou** [comment]
Yes, it's certainly fine to fix the tests so as to match the new (and better) error message.

**Review by pitrou — APPROVED** [review]
Thanks a lot @benibus ! I rebased and will merge if there are no related CI failures.

**Pull request merged**